apiv6/kentikapi/client_integration_test.go

```diff
@@ -0,0 +1,369 @@
+		"type":       "global",
+		"os":         "I use Manjaro BTW",
+		"ip":         "95.179.136.58",
+		"lat":        52.374031,
+		"long":       4.88969,
+		"lastAuthed": "2020-07-09T21:37:00.826Z",
+		"family":     "IP_FAMILY_DUAL",
+		"asn":        20473.0,
+		"siteId":     "2137",
+		"version":    "0.0.2",
+		"challenge":  "dummy-challenge",
+		"city":       "Amsterdam",
+		"region":     "Noord-Holland",
+		"country":    "Netherlands",
+		"testIds":    []interface{}{"13", "133", "1337"},
+		"localIp":    "10.10.10.10",
+	}
+}
+
+const dummyAgentResponseBody string = `{
+	"agent": {
+		"id": "968",
+		"name": "dummy-agent",
+		"status": "AGENT_STATUS_WAIT",
+		"alias": "probe-4-ams-1",
+		"type": "global",
+		"os": "I use Manjaro BTW",
+		"ip": "95.179.136.58",
+		"lat": 52.374031,
+		"long": 4.88969,
+		"lastAuthed": "2020-07-09T21:37:00.826Z",
+		"family": "IP_FAMILY_DUAL",
+		"asn": 20473,
+		"siteId": "2137",
+		"version": "0.0.2",
+		"challenge": "dummy-challenge",
+		"city": "Amsterdam",
+		"region": "Noord-Holland",
+		"country": "Netherlands",
+		"testIds": [
+			"13",
+			"133",
+			"1337"
+		],
+		"localIp": "10.10.10.10"
+	}
+}`
+
+type spyHTTPHandler struct {
+	t testing.TB
+	// responses to return to the client
+	responses []httpResponse
+
+	// requests spied by the handler
+	requests []httpRequest
+}
+
+func newSpyHTTPHandler(t testing.TB, responses []httpResponse) *spyHTTPHandler {
+	return &spyHTTPHandler{
+		t:         t,
+		responses: responses,
+	}
+}
+
+func (h *spyHTTPHandler) ServeHTTP(rw http.ResponseWriter, r *http.Request) {
+	body, err := ioutil.ReadAll(r.Body)
+	assert.NoError(h.t, err)
+
+	err = r.Body.Close()
+	assert.NoError(h.t, err)
+
+	h.requests = append(h.requests, httpRequest{
+		method: r.Method,
+		url_:   r.URL,
+		header: r.Header,
+		body:   string(body),
+	})
+
+	rw.Header().Set("Content-Type", "application/json")
+	response := h.response()
+	rw.WriteHeader(response.statusCode)
+	_, err = rw.Write([]byte(response.body))
+	assert.NoError(h.t, err)
+}
+
+func (h *spyHTTPHandler) response() httpResponse {
+	if len(h.requests) > len(h.responses) {
+		return httpResponse{
+			statusCode: http.StatusGone,
+			body: fmt.Sprintf(
+				"spyHTTPHandler: unexpected request, requests count: %v, expected: %v",
+				len(h.requests), len(h.responses),
+			),
+		}
+	}
+
+	return h.responses[len(h.requests)-1]
+}
+
+type httpRequest struct {
+	method string
+	url_   *url.URL
+	header http.Header
+	body   string
+}
+
+type httpResponse struct {
+	statusCode int
+	body       string
+}
+
+func newErrorHTTPResponse(statusCode int) httpResponse {
+	return httpResponse{
+		statusCode: statusCode,
+		body:       fmt.Sprintf(`{"error":"%v"}`, http.StatusText(statusCode)),
+	}
+}
+
+func unmarshalJSONToIf(t testing.TB, jsonString string) interface{} {
+	var data interface{}
+	err := json.Unmarshal([]byte(jsonString), &data)
+	assert.NoError(t, err)
+	return data
+}
+
+func stringPtr(s string) *string {
```

Comment:
Out of curiosity, why are all those (basically) typecast functions needed?

Reply:
Because string literals are not addressable. Nice write-up about it: https://stackoverflow.com/a/30716481/5000820